File: src/banks/manulife-bank/schemas.ts
import { tz } from "@date-fns/tz";
import { formatISO } from "date-fns";
import { v5 as uuidv5 } from "uuid";
import { z } from "zod";
import env from "../../utils/env.js";

const Account = z
  .object({
    id: z.string(),
    displayName: z.string(),
    accountId: z.object({
      accountNumber: z.string(),
    }),
    balance: z.number(),
  })
  .transform((account) => ({
    id: uuidv5(account.accountId.accountNumber, env.UUID_NAMESPACE),
    name: `${account.displayName} (${account.accountId.accountNumber})`,
    balance: account.balance,
    _index: account.id,
  }));

const AccountResponse = z
  .object({
    assetAccounts: z.object({
      assetAccount: z.array(Account),
    }),
  })
  .transform(({ assetAccounts: { assetAccount: accounts } }) => accounts);

const Transaction = z
  .object({
    date: z.number(),
    description: z.string(),
    transactionAmount: z.number(),
  })
  .transform((transaction) => ({
    date: formatISO(new Date(transaction.date), {
      representation: "date",
      in: tz(env.TZ),
    }),
    amount: transaction.transactionAmount,
    description: transaction.description,
  }));

const TransactionsResponse = z
  .object({
    historyTransactions: z.object({
      transaction: z.array(Transaction),
    }),
  })
  .transform(
    ({ historyTransactions: { transaction: transactions } }) => transactions,
  );

export { Account, AccountResponse, TransactionsResponse };

File: src/banks/manulife-bank/schemas.py
from dateutil import tz
from datetime import datetime
from uuid import uuid5
from pydantic import BaseModel, Field
import env

class Account(BaseModel):
    id: str
    displayName: str
    accountId: dict
    balance: float

    def transform(self):
        return {
            "id": str(uuid5(env.UUID_NAMESPACE, self.accountId["accountNumber"])),
            "name": f"{self.displayName} ({self.accountId['accountNumber']})",
            "balance": self.balance,
            "_index": self.id,
        }

class AccountResponse(BaseModel):
    assetAccounts: dict

    def transform(self):
        accounts = self.assetAccounts["assetAccount"]
        return [Account(**account).transform() for account in accounts]

class Transaction(BaseModel):
    date: int
    description: str
    transactionAmount: float

    def transform(self):
        date = datetime.fromtimestamp(self.date / 1000, tz=tz.gettz(env.TZ))
        return {
            "date": date.isoformat(),
            "amount": self.transactionAmount,
            "description": self.description,
        }

class TransactionsResponse(BaseModel):
    historyTransactions: dict

    def transform(self):
        transactions = self.historyTransactions["transaction"]
        return [Transaction(**transaction).transform() for transaction in transactions]
